### PR TITLE
Read HW version

### PR DESCRIPTION
**Pull request by yangtechen**
Read HW version from GPIO pin. If the pin reads low, it is P2. Otherwise, it is P1.

### CODE REVIEW

**Comment by yangtechen** [comment]
@eleenest After you make sure the logic works on P1, we can merge this branch.

File: source/hic_hal/stm32/stm32h743ii/read_ver.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include "IO_Config.h"
+
+/*
+ * Read UDB version from GPIO PG15 with weak pull-up register enabled
+ * Prerequisite: initialize GPIO PG15 at gpio_init() in gpio.c
+ */
+
+static char udb_board_version[4];
+
+void read_udb_board_version(void)
```
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/read_ver.c]: Can you put this file into the extended features folder? This is not something that Daplink or a standard stm32h743 hic does by default, it's just something that we are adding to our UDB project for our use case.

Maybe you could also call the file "udb_version.c" to be more descriptive. Then you could add other udb version related logic here and also change function names to read/get_udb_hw_version() to align with the file name.

File: source/hic_hal/stm32/stm32h743ii/read_ver.c [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  GPIO_PinState bitstatus;
+  uint8_t prefix_len = 0;
+  const char *prefix = "p";
+
+  prefix_len = strlen(prefix);
+  memset(udb_board_version, 0, sizeof(udb_board_version));
+  strcat(udb_board_version, prefix);
+  bitstatus = HAL_GPIO_ReadPin(PIN_UDB_VERSION_PORT, PIN_UDB_VERSION);
+
+  if (bitstatus == GPIO_PIN_RESET) {
```
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/read_ver.c]: In the Nest codebase, we always put the bracket on a newline. Probably it's good to build the habit early.

File: source/hic_hal/stm32/stm32h743ii/IO_Config.h [hunk 3]
```diff
@@ -119,6 +119,9 @@ COMPILER_ASSERT(DAPLINK_HIC_ID == DAPLINK_HIC_ID_STM32H743II);
 #define PIN_MSC_LED                  GPIO_PIN_4
 #define PIN_MSC_LED_Bit              4
 
+// UDB version (P1 or P2) detection
+#define PIN_UDB_VERSION_PORT         GPIOG
```
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/IO_Config.h]: My best understanding of the layers is this:
- Daplink (common layer that manages CMSIS, OS, and everything that's common between all Daplink projects)
- HIC (Hardware Interface Circuit, an implementation of the Daplink layer for a specific board, in our case these should be the drivers required for stm32h743 to run as a Daplink) 
- UDB (our superpowered Daplink stm32h743 HIC, UDB is a superset of the Daplink features)

If we want to contribute back to the original Daplink project, we have to keep the UDB features separate, since ARM probably doesn't want to maintain those. If another company wants to create a new Daplink HIC with the stm32h743 chip, we also don't want to force our custom features onto them, they might not want to read HW version on pin G5. 

Tldr; We probably want to create a UDB config layer in the extended features folder and keep UDB specific (not stm32h743 specific) configuration there.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c [hunk 4]
```diff
@@ -242,17 +244,22 @@ uint32_t DAP_ProcessVendorCommandEx(const uint8_t *request, uint8_t *response) {
     }
     case ID_DAP_VendorEx36_VERSION_DETAILS: {
         // Add a more specific internal version string.
-        static char build_version_str[] = "udb_" STR(UDB_VERSION) "_" GIT_COMMIT_SHA GIT_LOCAL_MODS_STR;
-        uint8_t len = strlen(build_version_str);
+        static char build_version_str[] = "udb_" STR(UDB_VERSION) "_" GIT_COMMIT_SHA GIT_LOCAL_MODS_STR "_";
```
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c]: Shouldn't we do all this once at boot instead of on each vendor command call?

We could keep all this logic in a udb_version.c file

File: source/hic_hal/stm32/stm32h743ii/read_ver.c [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+#include <string.h>
+#include "read_ver.h"
+#include "IO_Config.h"
+
+/*
+ * Read UDB version from GPIO PG15 with weak pull-up register enabled
+ * Prerequisite: initialize GPIO PG15 at gpio_init() in gpio.c
+ */
+
+static char udb_board_version[4];
```
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/read_ver.c]: In the Nest codebase, we prefix static variables with "s_"

File: source/hic_hal/stm32/stm32h743ii/IO_Config.h (hunk 3 above)
Comment (gaborcsapo) [reply source/hic_hal/stm32/stm32h743ii/IO_Config.h]: A lot more work is required to fully achieve this kind of separation, we should definitely keep new code to this standard and we can improve the old code when we get time to it.

**Comment by eleenest** [comment]
I checked out the branch, built it and tried it out.  

I tried testing the command on a p2 board, and two p1 boards.  
It looks like all 3 returned _p2 though.  
I checked with a DMM, the p2 board had about 9.5 kohm to gnd on that pin (available as pin 16 on the extended connector).  The two older boards had high impedance (as expected).  

With a DMM I measured low voltage on the adc1_2_3_exti15 pin for all 3 boards, so maybe the internal pullup isn't on?  That doesn't seem "as expected".
They "appeared" to be high-impedance (i.e. when touching the probe I could see the voltage change some, so didn't appear to be grounded, etc.)

I also tried jumpering over 1.8V (from pin 1 of the external connector) to pin 16, to "force" the pin high, like a p1 is expected to be.  I also saw it report back as a _p2 in the version command in this case also, though.  

To verify: 
* is the weak pullup enabled on the pin? (and the clock enabled to that bank of gpio's, pin configured for digital input (vs default of "analog mode"), etc.)
I'm guessing the pin might still be in "analog mode", so no weak pullup, and not really reading back the value from the pin.  


RESULTS:
newer board
pyocd> $ probe._link.vendor(36, [0x00]).tostring()
b'5udb_0.9.0_9b0facb72dea0206e49f666a13694872d503fdb6_p2'




older board
Connected to CoreSightTarget [no init mode]: 000000810046003b3139510d30303432a5a5a5a597969940
pyocd> $ probe._link.vendor(36, [0x00]).tostring()
b'5udb_0.9.0_9b0facb72dea0206e49f666a13694872d503fdb6_p2'


Connected to CoreSightTarget [no init mode]: 00000081004d002b3030510937393435a5a5a5a597969940
pyocd> $ probe._link.vendor(36, [0x00]).tostring()
b'5udb_0.9.0_9b0facb72dea0206e49f666a13694872d503fdb6_p2'
pyocd>

**Comment by yangtechen** [comment]
Apologies for not testing it thoroughly.

I printed the GPIOG PUPDR with gdb and got into the following weird situation.
There are two possible values of PUPDR before entering read_udb_version()
If the value of PUPDR is 0x2d57fd7d, I can set pull-up register correctly and the PUPDR will be 0x2d57f57d.
(Since I use pin 5 to read hw version and pull-up is 01 in the manual. so I think "d (1101) -> 5 (0101)" implies that I set it correctly.)

But if the value of PUPDR is 760741245, the it will be set to 760739197 by the code. This can not be right since it doesn't change the PUPDR of pin 5 but change other pins.

Unfortunately, the PUPDR value is wrong most of the time.

**Comment by yangtechen** [comment]
I printed the PUPDR value of GPIOG pin5 and it is set to 01 (pull-up) as expected.
And the MODER value is 00 (input).
But I am not sure if this is enough to indicate the pull-up is on.

**Comment by yangtechen** [comment]
Sorry I read the wrong pin. I think it is a typo so that I read pin 5 not pin15.

**Comment by eleenest** [comment]
I tested again.  

####
On a p2:

 $ probe._link.vendor(36, [0x00]).tostring()
b'5udb_0.9.0_fc039145c143fc204826db582cc1093df6567fd5_p2'

####

on a p1

$ probe._link.vendor(36, [0x00]).tostring()
b'5udb_0.9.0_fc039145c143fc204826db582cc1093df6567fd5_p1'

and the pin voltage was 1.78V
####
On a P2 the pin voltage was 0.467V, 
.467/1.78 = 0.262 which seems "about right" and very close to the ideal "0.259" ratio.  

####
So, this looks good to me.  
I'll add a comment to our "documentation bug", we might document this pin as being "used" (or needs special attention, external circuitry should not interfere with this measurement).

**Review by eleenest — APPROVED** [review]
Approving based on testing of the latest commit.
